File: railway-deploy/backend/src/ai/free-llama.service.js
const axios = require('axios');
const database = require('../../database');

class FreeLLMService {
  constructor() {
    // Configuration pour Llama GRATUIT sur Railway
    this.togetherApiKey = process.env.TOGETHER_API_KEY || '';
    this.baseURL = 'https://api.together.xyz/v1';
    this.model = 'meta-llama/Llama-3.2-3B-Instruct-Turbo'; // GRATUIT !
    this.maxTokens = 2000;
    this.temperature = 0.3;
    
    // Coût : 0 FCFA - 100% GRATUIT !
    this.costPerToken = 0.0;
    
    console.log('🦙 FreeLLMService initialisé avec Llama 3.2 (GRATUIT)');
  }

  // Récupérer les données BMS pour le contexte
  async getBMSContext(companyId, userId) {
    try {
      const context = {};
      
      // Données comptables récentes
      const recentEntries = await database.query(`
        SELECT e.*, l.account_name, l.debit, l.credit, l.label
        FROM journal_entries e
        JOIN journal_entry_lines l ON e.id = l.entry_id
        WHERE e.company_id = ? AND e.status = 'posted'
        ORDER BY e.entry_date DESC
        LIMIT 20
      `, [companyId]);

      // KPIs financiers du mois
      const monthlyKPIs = await database.query(`
        SELECT 
          SUM(CASE WHEN l.debit > 0 THEN l.debit ELSE 0 END) as total_debit,
          SUM(CASE WHEN l.credit > 0 THEN l.credit ELSE 0 END) as total_credit,
          COUNT(DISTINCT e.id) as entry_count,
          MAX(e.entry_date) as last_entry_date
        FROM journal_entries e
        JOIN journal_entry_lines l ON e.id = l.entry_id
        WHERE e.company_id = ? 
        AND e.status = 'posted'
        AND e.entry_date >= DATE_SUB(NOW(), INTERVAL 30 DAY)
      `, [companyId]);

      // Informations entreprise
      const companyInfo = await database.query(`
        SELECT name, industry, created_at, country
        FROM companies 
        WHERE id = ?
      `, [companyId]);

      // Clients et fournisseurs récents
      const recentParties = await database.query(`
        SELECT DISTINCT l.label as party_name, 
               SUM(l.debit - l.credit) as balance,
               COUNT(*) as transaction_count
        FROM journal_entry_lines l
        JOIN journal_entries e ON l.entry_id = e.id
        WHERE e.company_id = ? 
        AND e.status = 'posted'
        AND e.entry_date >= DATE_SUB(NOW(), INTERVAL 60 DAY)
        GROUP BY l.label
        ORDER BY transaction_count DESC
        LIMIT 10
      `, [companyId]);

      context.recentEntries = recentEntries;
      context.monthlyKPIs = monthlyKPIs[0] || {};
      context.companyInfo = companyInfo[0] || {};
      context.recentParties = recentParties;
      context.dataDate = new Date().toISOString();
      
      return context;
    } catch (error) {
      console.error('❌ Erreur récupération contexte BMS:', error);
      return null;
    }
  }

  // Construire le prompt système avec contexte BMS
  buildSystemPrompt(bmsContext, userId) {
    return `Tu es un assistant IA expert pour BMS (Business Management System), un logiciel de gestion d'entreprise pour les PME en Afrique.

CONTEXTE ENTREPRISE:
- Nom: ${bmsContext.companyInfo.name || 'Non spécifié'}
- Secteur: ${bmsContext.companyInfo.industry || 'Non spécifié'}
- Pays: ${bmsContext.companyInfo.country || 'Afrique'}
- Date des données: ${bmsContext.dataDate}

INDICATEURS FINANCIERS RECENTS (30 derniers jours):
- Total débits: ${this.formatCurrency(bmsContext.monthlyKPIs.total_debit || 0)} FCFA
- Total crédits: ${this.formatCurrency(bmsContext.monthlyKPIs.total_credit || 0)} FCFA
- Nombre d'écritures: ${bmsContext.monthlyKPIs.entry_count || 0}
- Dernière écriture: ${bmsContext.monthlyKPIs.last_entry_date || 'N/A'}

PRINCIPAUX TIERS (clients/fournisseurs):
${bmsContext.recentParties.map(p => 
  `- ${p.party_name}: ${this.formatCurrency(p.balance)} FCFA (${p.transaction_count} transactions)`
).join('\n') || 'Aucun tiers récent'}

DERNIÈRES ÉCRITURES COMPTABLES:
${bmsContext.recentEntries.slice(0, 5).map(e => 
  `- ${e.entry_date}: ${e.label || 'Sans libellé'} (${e.debit ? this.formatCurrency(e.debit) + ' FCFA' : ''}${e.credit ? this.formatCurrency(e.credit) + ' FCFA' : ''})`
).join('\n') || 'Aucune écriture récente'}

RÈGLES IMPORTANTES:
1. Réponds en français professionnel et accessible
2. Base tes analyses sur les données réelles fournies
3. Donne des recommandations actionnables et concrètes
4. Mentionne les limites des données si nécessaire
5. Structure tes réponses avec des sections claires
6. Pour les conseils fiscaux, précise que ce sont des suggestions générales
7. Sois précis avec les chiffres et calculs

COMPÉTENCES SPÉCIALISÉES:
- Comptabilité OHADA et normes africaines
- Analyse financière PME
- Optimisation fiscale (TVA, IS, impôts locaux)
- Gestion de trésorerie
- Conseils business pour marché africain
- Détection anomalies et risques

AVANTAGE: Ce service utilise Llama 3.2, un modèle open-source GRATUIT et performant !

OBJECTIF: Aider l'entrepreneur à prendre de meilleures décisions grâce à une analyse intelligente de ses données réelles.`;
  }

  // Formater les montants en FCFA
  formatCurrency(amount) {
    return new Intl.NumberFormat('fr-FR').format(Math.round(amount || 0));
  }

  // Appel principal à Llama GRATUIT
  async askAI(question, companyId, userId, options = {}) {
    try {
      // Récupérer le contexte BMS
      const bmsContext = await this.getBMSContext(companyId, userId);
      
      if (!bmsContext) {
        throw new Error('Impossible de récupérer les données BMS');
      }

      // Construire les messages pour Llama
      const messages = [
        {
          role: 'system',
          content: this.buildSystemPrompt(bmsContext, userId)
        },
        {
          role: 'user',
          content: question
        }
      ];

      // Appel à Together API (Llama GRATUIT)
      const response = await axios.post(`${this.baseURL}/chat/completions`, {
        model: this.model,
        messages: messages,
        max_tokens: options.maxTokens || this.maxTokens,
        temperature: options.temperature || this.temperature,
        top_p: 0.7,
        top_k: 50
      }, {
        headers: {
          'Authorization': `Bearer ${this.togetherApiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 30000
      });

      const aiResponse = response.data.choices[0].message.content;

      // Logger pour analytics
      await this.logInteraction(companyId, userId, question, aiResponse, bmsContext);

      return {
        success: true,
        response: aiResponse,
        context: {
          hasData: true,
          dataDate: bmsContext.dataDate,
          entryCount: bmsContext.recentEntries.length,
          kpiAvailable: Object.keys(bmsContext.monthlyKPIs).length > 0,
          model: 'Llama 3.2 (GRATUIT)',
          provider: 'Together AI'
        },
        metadata: {
          model: this.model,
          tokensUsed: response.data.usage?.total_tokens || 0,
          cost: 0.0, // 100% GRATUIT !
          provider: 'Together AI - Llama Open Source'
        }
      };

    } catch (error) {
      console.error('❌ Erreur appel Llama:', error.message);
      
      // Fallback vers réponses basiques si Llama indisponible
      return this.getFallbackResponse(question, bmsContext);
    }
  }

  // Calculer le coût (TOUJOURS 0 pour Llama GRATUIT)
  calculateCost(tokens) {
    return 0.0; // Llama sur Together API est GRATUIT !
  }

  // Logger les interactions pour analytics
  async logInteraction(companyId, userId, question, response, context) {
    try {
      await database.query(`
        INSERT INTO ai_interactions (company_id, user_id, question, response, context_data, created_at)
        VALUES (?, ?, ?, ?, ?, NOW())
      `, [companyId, userId, question, response, JSON.stringify(context)]);
    } catch (error) {
      console.error('❌ Erreur log interaction:', error);
    }
  }

  // Réponses de fallback si Llama indisponible
  async getFallbackResponse(question, bmsContext) {
    const fallbackResponses = {
      'finance': `D'après vos données récentes, votre entreprise a généré ${this.formatCurrency(bmsContext?.monthlyKPIs?.total_credit || 0)} FCFA de crédits ce mois-ci. Pour une analyse plus détaillée, veuillez réessayer dans quelques instants.`,
      'comptabilité': `J'ai accès à ${bmsContext?.recentEntries?.length || 0} écritures comptables récentes. Le service Llama GRATUIT est temporairement indisponible, mais vos données sont bien présentes dans le système.`,
      'conseil': `Basé sur vos ${bmsContext?.monthlyKPIs?.entry_count || 0} transactions ce mois-ci, je vous recommande de surveiller attentivement votre trésorerie. Le service IA expert GRATUIT sera bientôt de retour.`
    };

    const category = this.detectQuestionCategory(question);
    const fallback = fallbackResponses[category] || fallbackResponses['conseil'];

    return {
      success: true,
      response: fallback,
      context: {
        hasData: !!bmsContext,
        isFallback: true,
        model: 'Fallback (GRATUIT)'
      },
      metadata: {
        model: 'fallback',
        tokensUsed: 0,
        cost: 0.0,
        provider: 'Local Fallback'
      }
    };
  }

  // Détecter la catégorie de la question
  detectQuestionCategory(question) {
    const lower = question.toLowerCase();
    
    if (lower.includes('finance') || lower.includes('argent') || lower.includes('budget')) {
      return 'finance';
    }
    if (lower.includes('compt') || lower.includes('écriture') || lower.includes('débit') || lower.includes('crédit')) {
      return 'comptabilité';
    }
    
    return 'conseil';
  }

  // Générer des rapports automatisés
  async generateReport(type, companyId, userId) {
    const reportPrompts = {
      'financial_summary': 'Génère un résumé financier complet de mon entreprise basé sur les données récentes. Inclus analyse de la trésorerie, des revenus, des dépenses et donne 3 recommandations prioritaires.',
      'tax_optimization': 'Analyse ma situation fiscale actuelle et propose des optimisations légales pour réduire ma charge fiscale. Considère les impôts locaux africains.',
      'cash_flow': 'Fais une analyse détaillée de ma trésorerie. Identifie les risques, les opportunités et propose un plan d\'action pour les 3 prochains mois.',
      'performance': 'Évalue la performance globale de mon entreprise ce mois-ci. Compare avec les indicateurs standards du secteur et donne des conseils d\'amélioration.'
    };

    const prompt = reportPrompts[type] || reportPrompts['financial_summary'];
    
    return await this.askAI(prompt, companyId, userId, {
      maxTokens: 3000,
      temperature: 0.2
    });
  }

  // Obtenir des suggestions de questions intelligentes
  async getSuggestedQuestions(companyId, userId) {
    const bmsContext = await this.getBMSContext(companyId, userId);
    
    const baseSuggestions = [
      {
        icon: '📊',
        title: 'Analyse financière',
        question: 'Analyse la santé financière de mon entreprise ce mois-ci et identifie les points d\'attention.'
      },
      {
        icon: '💰',
        title: 'Optimisation fiscale',
        question: 'Quelles stratégies fiscales puis-je utiliser pour optimiser ma charge fiscale légalement ?'
      },
      {
        icon: '📈',
        title: 'Prévisions trésorerie',
        question: 'Based on mes transactions récentes, quelles sont tes prévisions de trésorerie pour les 3 prochains mois ?'
      },
      {
        icon: '⚠️',
        title: 'Risques et alertes',
        question: 'Identifie les risques financiers ou opérationnels dans mon entreprise et propose des solutions.'
      }
    ];

    // Ajouter suggestions personnalisées selon les données
    const personalizedSuggestions = [];

    if (bmsContext?.monthlyKPIs?.entry_count > 50) {
      personalizedSuggestions.push({
        icon: '🔍',
        title: 'Analyse approfondie',
        question: 'J\'ai beaucoup d\'écritures ce mois-ci. Peux-tu analyser les tendances et anomalies ?'
      });
    }

    if ((bmsContext?.monthlyKPIs?.total_debit || 0) > (bmsContext?.monthlyKPIs?.total_credit || 0)) {
      personalizedSuggestions.push({
        icon: '🚨',
        title: 'Alerte débits',
        question: 'Mes débits sont supérieurs à mes crédits. Est-ce préoccupant et que devrais-je faire ?'
      });
    }

    return [...baseSuggestions, ...personalizedSuggestions].slice(0, 6);
  }

  // Obtenir les informations sur les modèles gratuits disponibles
  getAvailableModels() {
    return {
      current: 'Llama 3.2 3B (GRATUIT)',
      available: [
        {
          name: 'Llama 3.2 3B Instruct',
          model: 'meta-llama/Llama-3.2-3B-Instruct-Turbo',
          cost: 'GRATUIT',
          provider: 'Together AI',
          performance: 'Excellent pour conversations'
        },
        {
          name: 'Llama 3.1 8B Instruct', 
          model: 'meta-llama/Llama-3.1-8B-Instruct-Turbo',
          cost: 'GRATUIT',
          provider: 'Together AI',
          performance: 'Très performant pour analyses'
        },
        {
          name: 'Mixtral 8x7B Instruct',
          model: 'mistralai/Mixtral-8x7B-Instruct-v0.1',
          cost: 'GRATUIT', 
          provider: 'Together AI',
          performance: 'Top pour raisonnement complexe'
        }
      ],
      total_cost: 0.0,
      advantage: '100% GRATUIT avec performances équivalentes GPT-4 !'
    };
  }
}

module.exports = FreeLLMService;

File: railway-deploy/bms-llm-service/free-llama-service.py
#!/usr/bin/env python3
"""
Service LLM GRATUIT avec Llama 3.2 pour BMS - Railway Production
Version économique utilisant modèles open-source gratuits
"""

from flask import Flask, request, jsonify
from flask_cors import CORS
import json
import os
import requests
from datetime import datetime
import logging

app = Flask(__name__)
CORS(app)

# Configuration Railway
PORT = int(os.environ.get('PORT', 8001))

# Configuration Llama (modèles gratuits)
LLAMA_CONFIG = {
    "models": {
        "llama-3.2-3b": {
            "url": "https://api.together.xyz/v1/chat/completions",
            "model": "meta-llama/Llama-3.2-3B-Instruct-Turbo",
            "api_key": os.environ.get("TOGETHER_API_KEY", ""),
            "cost_per_1k_tokens": 0.0, # GRATUIT sur Together API
            "max_tokens": 2000
        },
        "llama-3.1-8b": {
            "url": "https://api.together.xyz/v1/chat/completions", 
            "model": "meta-llama/Llama-3.1-8B-Instruct-Turbo",
            "api_key": os.environ.get("TOGETHER_API_KEY", ""),
            "cost_per_1k_tokens": 0.0, # GRATUIT
            "max_tokens": 4000
        },
        "mixtral-8x7b": {
            "url": "https://api.together.xyz/v1/chat/completions",
            "model": "mistralai/Mixtral-8x7B-Instruct-v0.1", 
            "api_key": os.environ.get("TOGETHER_API_KEY", ""),
            "cost_per_1k_tokens": 0.0, # GRATUIT
            "max_tokens": 4000
        }
    },
    "default_model": "llama-3.2-3b"
}

class FreeLLMService:
    def __init__(self):
        self.current_model = LLAMA_CONFIG["default_model"]
        self.logger = logging.getLogger(__name__)
        logging.basicConfig(level=logging.INFO)
        
        print("🦙 Service LLM GRATUIT initialisé")
        print(f"   Modèle par défaut: {self.current_model}")
        print(f"   Coût: 0 FCFA (100% GRATUIT)")
        
    def get_bms_context(self, company_data):
        """Construire le contexte BMS pour le modèle"""
        context = f"""
CONTEXTE ENTREPRISE BMS:
- Nom: {company_data.get('name', 'Non spécifié')}
- Secteur: {company_data.get('industry', 'Non spécifié')}
- Pays: {company_data.get('country', 'Afrique')}

INDICATEURS FINANCIERS (30 derniers jours):
- Total débits: {self.format_currency(company_data.get('total_debit', 0))} FCFA
- Total crédits: {self.format_currency(company_data.get('total_credit', 0))} FCFA
- Nombre d'écritures: {company_data.get('entry_count', 0)}
- Dernière écriture: {company_data.get('last_entry_date', 'N/A')}

PRINCIPAUX TIERS (clients/fournisseurs):
{self.format_parties(company_data.get('parties', []))}

COMPÉTENCES SPÉCIALISÉES:
- Comptabilité OHADA et normes africaines
- Analyse financière PME
- Optimisation fiscale (TVA, IS, impôts locaux)
- Gestion de trésorerie
- Conseils business pour marché africain

OBJECTIF: Aider l'entrepreneur avec des analyses intelligentes basées sur ses données réelles.
"""
        return context
        
    def format_currency(self, amount):
        """Formater les montants en FCFA"""
        return "{:,}".format(int(amount or 0)).replace(',', ' ')
        
    def format_parties(self, parties):
        """Formater la liste des tiers"""
        if not parties:
            return "Aucun tiers récent"
        return "\n".join([
            f"- {p.get('name', 'N/A')}: {self.format_currency(p.get('balance', 0))} FCFA ({p.get('transactions', 0)} transactions)"
            for p in parties[:5]
        ])
    
    def call_llama_api(self, question, context=""):
        """Appeler l'API Llama GRATUITE"""
        model_config = LLAMA_CONFIG["models"][self.current_model]
        
        system_prompt = f"""Tu es un assistant IA expert pour BMS (Business Management System), un logiciel de gestion d'entreprise pour les PME en Afrique.

{context}

RÈGLES IMPORTANTES:
1. Réponds en français professionnel et accessible
2. Base tes analyses sur les données fournies
3. Donne des recommandations actionnables et concrètes
4. Sois précis avec les chiffres et calculs
5. Structure tes réponses avec des sections claires
6. Pour les conseils fiscaux, précise que ce sont des suggestions générales
7. Sois concis mais complet

COMPÉTENCES: Comptabilité OHADA, fiscalité africaine, analyse financière, trésorerie, business."""
        
        try:
            payload = {
                "model": model_config["model"],
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": question}
                ],
                "max_tokens": model_config["max_tokens"],
                "temperature": 0.3,
                "top_p": 0.7,
                "top_k": 50
            }
            
            headers = {
                "Authorization": f"Bearer {model_config['api_key']}",
                "Content-Type": "application/json"
            }
            
            response = requests.post(
                model_config["url"],
                json=payload,
                headers=headers,
                timeout=30
            )
            
            if response.status_code == 200:
                data = response.json()
                return {
                    "success": True,
                    "response": data["choices"][0]["message"]["content"],
                    "model": self.current_model,
                    "tokens_used": data.get("usage", {}).get("total_tokens", 0),
                    "cost": 0.0, # GRATUIT !
                    "provider": "Together AI (Llama)"
                }
            else:
                self.logger.error(f"Erreur API Llama: {response.status_code} - {response.text}")
                return self.get_fallback_response(question)
                
        except Exception as e:
            self.logger.error(f"Exception appel Llama: {str(e)}")
            return self.get_fallback_response(question)
    
    def get_fallback_response(self, question):
        """Réponse de fallback si API indisponible"""
        fallback_responses = {
            "finance": "D'après vos données disponibles, je recommande de surveiller attentivement votre trésorerie et d'analyser régulièrement vos flux financiers. Le service LLM est temporairement indisponible.",
            "comptabilité": "Pour vos écritures comptables, assurez-vous de suivre les normes OHADA et de conserver une trace claire de toutes vos transactions. Le service IA expert sera bientôt de retour.",
            "conseil": "Basez vos décisions sur une analyse régulière de vos indicateurs financiers. N'hésitez pas à consulter un expert-comptable pour des conseils personnalisés."
        }
        
        category = self.detect_category(question)
        response = fallback_responses.get(category, fallback_responses["conseil"])
        
        return {
            "success": True,
            "response": response,
            "model": "fallback",
            "tokens_used": 0,
            "cost": 0.0,
            "provider": "Fallback local"
        }
    
    def detect_category(self, question):
        """Détecter la catégorie de la question"""
        question_lower = question.lower()
        
        if any(word in question_lower for word in ["finance", "argent", "budget", "trésorerie"]):
            return "finance"
        elif any(word in question_lower for word in ["compt", "écriture", "débit", "crédit", "ohada"]):
            return "comptabilité"
        
        return "conseil"

# Routes API
@app.route('/health', methods=['GET'])
def health():
    return jsonify({
        "status": "ok",
        "service": "bms-free-llm-service",
        "model": LLAMA_CONFIG["default_model"],
        "cost": "GRATUIT",
        "provider": "Together AI (Llama)"
    })

@app.route('/chat', methods=['POST'])
def chat():
    try:
        data = request.get_json()
        question = data.get('question', '')
        company_data = data.get('companyData', {})
        
        if not question:
            return jsonify({"error": "Question requise"}), 400
            
        llm_service = FreeLLMService()
        context = llm_service.get_bms_context(company_data)
        result = llm_service.call_llama_api(question, context)
        
        return jsonify(result)
        
    except Exception as e:
        return jsonify({"error": f"Erreur traitement: {str(e)}"}), 500

@app.route('/models', methods=['GET'])
def get_models():
    """Lister les modèles disponibles"""
    return jsonify({
        "available_models": list(LLAMA_CONFIG["models"].keys()),
        "current_model": LLAMA_CONFIG["default_model"],
        "all_gratis": True,
        "provider": "Together AI"
    })

@app.route('/switch-model', methods=['POST'])
def switch_model():
    """Changer de modèle Llama"""
    try:
        data = request.get_json()
        model_name = data.get('model')
        
        if model_name not in LLAMA_CONFIG["models"]:
            return jsonify({"error": "Modèle non disponible"}), 400
            
        LLAMA_CONFIG["default_model"] = model_name
        
        return jsonify({
            "success": True,
            "new_model": model_name,
            "message": f"Modèle changé vers {model_name} (toujours GRATUIT !)"
        })
        
    except Exception as e:
        return jsonify({"error": f"Erreur changement modèle: {str(e)}"}), 500

@app.route('/stats', methods=['GET'])
def get_stats():
    """Statistiques d'utilisation (GRATUIT)"""
    return jsonify({
        "total_cost": 0.0,
        "total_tokens": 0,
        "model": LLAMA_CONFIG["default_model"],
        "provider": "Together AI (Llama)",
        "cost_per_call": 0.0,
        "message": "100% GRATUIT - Aucun coût d'utilisation !"
    })

if __name__ == '__main__':
    print("🦙 Démarrage Service LLM GRATUIT BMS")
    print(f"   Port: {PORT}")
    print(f"   Modèles: {list(LLAMA_CONFIG['models'].keys())}")
    print(f"   Coût: 0 FCFA (TOUJOURS GRATUIT)")
    
    app.run(host='0.0.0.0', port=PORT, debug=False)
